### DISCUSSION TITLE
max_requiv problem

### DISCUSSION THREAD

**Discussion by MohammadReza0GH** [phoebe-project/phoebe2#855] [Questions]
hi im trying to fit a data (semidetached) , bot i got this problem:

secondary is overflowing at periastron (requiv=1.0, requiv_max=0.0835699129883548).  Use contact model if overflowing is desired. (3 affected parameters, affecting system,run_compute) when requiv of primary =0.0835699129883548=requiv_max, why they are always same?? i changed parameters but this happened again. i did : b.add_constraint('semidetached','primary')

this is my code:
import numpy as np
import phoebe

b = phoebe.default_binary()
logger = phoebe.logger(clevel = 'WARNING')
lc = np.loadtxt('D:\J2215.txt')

fluxes = np.empty(81)
for i in range(81):
    fluxes[i] = 1373*10**((-26-lc[i,1])/2.5)
    
b.add_constraint('semidetached','primary')
b.set_value('requiv', component='secondary', value=0.00014)
b.set_value('eclipse_method', value ='only_horizon')
b.set_value('sma', component = 'binary', value= 0.22)
b.set_value_all('atm',  value='blackbody')
b.set_value_all('ld_mode_bol',  value='manual')
b.set_value('distance@system', value = 150, unit = 'pc')
times = b.to_time(lc[:,0])
b.add_dataset('lc', times = times, fluxes = fluxes, passband = 'Johnson:V',ld_mode='manual',pblum_mode='absolute', dataset= 'lc01')
b.set_value_all('ld_mode', value='manual')

b.run_compute()
b.add_solver('estimator.lc_periodogram', solver='lcperiod', algorithm = 'ls')
b.run_solver('lcperiod', solution='lcperiod_sol')
b.adopt_solution('lcperiod_sol')

b.add_solver('estimator.ebai', solver = 'ebai_1')
b.run_solver('ebai_1', solution='ebai1_sol')
#b.add_constraint('requivsumfrac')
#b.add_constraint('requivratio')
#b.flip_constraint('requivratio', solve_for='sma@binary')
b.flip_constraint('requivsumfrac', solve_for='requiv@secondary')
#b.add_constraint('teffratio')
b.flip_constraint('teffratio', solve_for='teff@secondary')
b.flip_constraint('ecosw', solve_for='per0')
b.flip_constraint('esinw', solve_for='ecc')
b.adopt_solution('ebai1_sol')

b.add_solver('estimator.lc_geometry', solver = 'lcgeo',eclipse_method ='only_horizon', overwrite=True)
b.run_solver('lcgeo', solution = 'lcgeo_sol',overwrite=True)
for param,value,unit in zip(b.get_value('fitted_twigs', solution='lcgeo_sol'),b.get_value('fitted_values', solution = 'lcgeo_sol'), b.get_value('fitted_units', solution = 'lcgeo_sol')):
    print((param,value,unit))

b.flip_constraint('per0', solve_for='ecosw')
b.flip_constraint('ecc', solve_for='esinw')
b.adopt_solution('lcgeo_sol')

b.run_compute(model = 'ebai_model')
b.run_compute(model = 'lcgeom_model')
b.plot(x='phases', legend = True, show=True)

**Comment by MohammadReza0GH** [comment]
[J2215.txt](https://github.com/phoebe-project/phoebe2/files/15110691/J2215.txt)

this is the observation data that Im using.

**Comment by kecnry** [comment]
`b.add_constraint('semidetached','primary')` tells the primary to _always_ have `requiv == requiv_max`, by definition (see [this tutorial on the semi-detached constraint](http://phoebe-project.org/docs/latests/tutorials/requiv_crit_semidetached)).  Alternatively, you can omit the constraint and set the radius to at or near the critical value manually, the constraint just enforces the radius to update as you update other parameters that affect the critical value.

The error message, however, is telling you that the other star is overflowing its critical radius.  If this is just showing as a warning (through the logger), then it might just be intermediate and you fix the problem later in the script.  If it is being raised as an error when you call `run_compute`, then that means that your system is not physical and the parameters need to be adjusted before the model can be calculated - in which case you'll need to decrease the radius of the secondary star (and/or increase the semi-major axis, and/or decrease eccentricity, etc.)

**Reply by MohammadReza0GH** [reply]
yes,  b.add_constraint('semidetached','primary') tells the primary to always have requiv == requiv_max.

but i got  requiv@primary==requiv_max@secondary so this is a problem, and i cheanged parameters that you mentioned this happens again. like :
requiv@secondary=7.0213591564273905 and requiv_max@secondary=2.8013682812124276 but requiv@primary=2.8013682812124276

[requiv@primary==requiv_max@secondary ] => this is wrong, isnt it?

**Reply by kecnry** [reply]
Ok, sorry if I misunderstood.  Can you please post the output of `print(b.filter(qualifier='requiv*'))`?

**Reply by MohammadReza0GH** [reply]
requiv@primary@component: 2.8013682812124276 solRad
C    requiv_max@primary@component: 2.8013682812124276 solRad
C      requiv@secondary@component: 7.0213591564273905 solRad
C  requiv_max@secondary@component: 2.8013682812124276 solRad
C    requivratio@binary@component: 2.5064034613073285
   requivsumfrac@binary@component: 0.9822727437639818
           requivratio@constraint: {requiv@secondary@component} / {requiv@primary@component}
      requiv@secondary@constraint: ({requivsumfrac@binary@component} * {sma@binary@component}) - {requiv@primary@component}
    requiv_max@primary@constraint: requiv_L1({q@binary@component}, {syncpar@primary@component}, {ecc@binary@component}, {sma@binary@component}, {incl@primary@component}, {long_an@primary@component}, {incl@binary@component}, {long_an@binary@component}, 1)
   requiv_max@secondary@constr...: requiv_L1({q@binary@component}, {syncpar@secondary@component}, {ecc@binary@component}, {sma@binary@component}, {incl@secondary@component}, {long_an@secondary@component}, {incl@binary@component}, {long_an@binary@component}, 2)
        requiv@primary@constraint: 1.000000 * {requiv_max@primary@component}

**Reply by kecnry** [reply]
ok, that definitely doesn't look right 🤔.  My guess is that one of the constraint flips caused a problem which probably suggests a bug, but one that we can probably workaround in the meantime.  Can you go through your script, add this same print statement before and after any `flip_constraint` or `adopt_solution` lines and try to pinpoint at what point the first line (`requiv@primary@component`) lost the "C" (which indicates that it is constrained)?

**Reply by MohammadReza0GH** [reply]
first line dose not lose C that was my copy/paste mistake sorry.

before b.add_constraint('semidetached','primary') : 
         requiv@primary@component: 1.0 solRad  <<<<
C    requiv_max@primary@component: 2.0132751765376384 solRad
       requiv@secondary@component: 1.0 solRad
C  requiv_max@secondary@component: 2.0132751765376384 solRad  <<<<
C    requivratio@binary@component: 1.0
C  requivsumfrac@binary@component: 0.37735849056603776

after b.add_constraint('semidetached','primary') : 
C requiv@primary@component: 2.0132751765376384 solRad   <<<<
C    requiv_max@primary@component: 2.0132751765376384 solRad
       requiv@secondary@component: 1.0 solRad
C  requiv_max@secondary@component: 2.0132751765376384 solRad    <<<<  
C    requivratio@binary@component: 0.4967030894007076
C  requivsumfrac@binary@component: 0.5685424861391771

after flips:
C        requiv@primary@component: 3.798632408561582 solRad
C    requiv_max@primary@component: 3.798632408561582 solRad
C      requiv@secondary@component: 0.000140000000000029 solRad
C  requiv_max@secondary@component: 3.798632408561582 solRad
C    requivratio@binary@component: 3.68553692335325e-05
   requivsumfrac@binary@component: 0.3798772408561582

after adopt_solution:
C requiv@primary@component: 2.8013682812124276 solRad
C    requiv_max@primary@component: 2.8013682812124276 solRad
C      requiv@secondary@component: 7.0213591564273905 solRad
C  requiv_max@secondary@component: 2.8013682812124276 solRad
C    requivratio@binary@component: 2.5064034613073285
   requivsumfrac@binary@component: 0.9822727437639818

**Reply by kecnry** [reply]
oh ok, so then I think everything is okay.  The `requiv_max` are the same for the primary and secondary because the mass ratio `q` is one.  You can see what goes in to computing `requiv_max` in the constraint definitions: 

```
requiv_max@primary@constraint: requiv_L1({q@binary@component}, {syncpar@primary@component}, {ecc@binary@component}, {sma@binary@component}, {incl@primary@component}, {long_an@primary@component}, {incl@binary@component}, {long_an@binary@component}, 1)
```

**Reply by MohammadReza0GH** [reply]
thanks.

**Reply by MohammadReza0GH** [reply]
i tried to fit with only EBAI but i got this:
![hsrtghsrvbatngewrGBavsrdf](https://github.com/phoebe-project/phoebe2/assets/167889059/98425f0c-7aed-4890-b2ad-8bb04549b107)

**Reply by kecnry** [reply]
EBAI, along with the other estimators, acts on phased data and proposes some values as an estimate (but should not be expected to reproduce the light curve immediately or even work well on all cases).  Unless the orbital period of this system is actually ~1, you'll need to unphase the input data first before computing a physical model.  And either way, make sure that when plotted in phase space (`x=phases`) that the phased light curve looks correct to pass off to the estimator.

There are a number of examples and tutorials on the website (including some recordings from [past workshops](http://phoebe-project.org/workshops)) that you might also find helpful:
* [section 8 of the tutorials](http://phoebe-project.org/docs/latest/tutorials) is dedicated to the inverse problem (fitting)
* [inverse problem examples](http://phoebe-project.org/docs/latest/examples/inverse_paper_examples)

**Reply by MohammadReza0GH** [reply]
ok tnx
